File: m9/provas/p1/Python/publisher.py
import paho.mqtt.client as mqtt
import time
import json
import random
from faker import Faker

client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, "python_publisher")

client.connect("localhost", 1891, 60)

tipo = ['freezer', 'geladeira']
loja = [1, 2, 3, 4, 5, 6]
def SensorData():
    fake = Faker()
    sensor_register = {
        "Loja": random.choice(loja),
        "tipo": random.choice(tipo),
        "temperatura": fake.pyfloat(min_value=-25, max_value=25, right_digits=2),
        "hora": time.strftime("%Y-%m-%d %H:%M:%S")
    }
    return sensor_register


try:
    while True:
        data = SensorData()
        message = json.dumps(data)
        client.publish("test/topic", message)
        print(data['temperatura'])
        if -15> float(data['temperatura']) >-25 and data['tipo'] == 'freezer':
            print("Alerta, a temperatura no freezer está muito alta")
        if 2> float(data['temperatura']) >10 and data['tipo'] == 'geladeira':
            print("Alerta, a temperatura na geladeira está muito alta")
        print(f"Publicado: {message}")


        time.sleep(2)
except KeyboardInterrupt:
    print("Publicação encerrada")


client.disconnect()

File: m9/provas/p1/GO/publisher.go
package main

import (
	"encoding/json"
	"flag"
	"log"
	"math/rand"
	"os"
	"os/signal"
	"syscall"
	"time"


	MQTT "github.com/eclipse/paho.mqtt.golang"
)


const (
	defaultBrokerURL = "tcp://localhost:1883"
	defaultClientID  = "publisher"
	topic            = "/pratica_p1"
)

 


func SensorData() map[string]int {
	data := map[string]int{
		
		"NH3_ppm": 
		"CO_ppm":  rand.Intn(1000),
		"NO2_ppm": rand.Intn(30),
	}
	return data
}

func Connect(brokerURL, clientID string) MQTT.Client {
	opts := MQTT.NewClientOptions().AddBroker(brokerURL)
	opts.SetClientID(clientID)

	client := MQTT.NewClient(opts)

	if token := client.Connect(); token.Wait() && token.Error() != nil {
		log.Fatalf("Error connecting to broker: %v", token.Error())
	}

	return client
}

func Publish(client MQTT.Client) {
	for {
		data := SensorData()
		jsonData, err := json.Marshal(data)
		if err != nil {
			log.Printf("Error converting data to JSON: %v", err)
			continue
		}

		msg := time.Now().Format(time.RFC3339) + " " + string(jsonData)

		token := client.Publish(topic, 1, false, msg)
		token.Wait()

		log.Printf("[PUBLISHER] %s", msg)
		time.Sleep(2 * time.Second)
	}
}

func main() {
	brokerURL := flag.String("broker", defaultBrokerURL, "MQTT broker URL")
	clientID := flag.String("clientID", defaultClientID, "Client ID")
	flag.Parse()

	client := Connect(*brokerURL, *clientID)
	defer client.Disconnect(250)

	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)

	go Publish(client)

	<-quit
	log.Println("Exiting...")
}